Modules/_datetimemodule.c

```diff
@@ -3596,10 +3595,9 @@ static PyMethodDef date_methods[] = {
      PyDoc_STR("Return the day of the week represented by the date.\n"
                "Monday == 0 ... Sunday == 6")},
 
-    {"replace",     _PyCFunction_CAST(date_replace),      METH_VARARGS | METH_KEYWORDS,
-     PyDoc_STR("Return date with new specified fields.")},
+    DATETIME_DATE_REPLACE_METHODDEF
 
-    {"__replace__", _PyCFunction_CAST(date_replace),      METH_VARARGS | METH_KEYWORDS},
+    {"__replace__", _PyCFunction_CAST(datetime_date_replace), METH_FASTCALL | METH_KEYWORDS},
```

Comment:
There is an issue here. DATETIME_DATE_REPLACE_METHODDEF can be changed with different options of Argument Clinic (for example when we experiment with the limited C API). The generated `datetime_date_replace` can no longer be compatible with manually written entry for `__replace__`.

The simplest way to fix this is to add separate definition for `__replace__`:
```c
/*[clinic input]
datetime.date.__replace__ = datetime.date.replace
[clinic start generated code]*/

static PyObject *
datetime_date___replace___impl(PyDateTime_Date *self, int year, int month,
                               int day)
/*[clinic end generated code: output=bfb97f926e24f201 input=1b2cac755dfa9b40]*/
{
    return new_date_ex(year, month, day, Py_TYPE(self));
}
```

Reply:
If I understand correctly, this means that the bodies of all replace() functions need to be duplicated. This means that all future changes to that code will need to be done twice, which is a bad thing. (The functions may seem trivial, but I'm already working on a further change to address #89039.) This will also generate the argument parsing code twice, which will bloat executable size. A better approach would be to add an option to argument clinic to generate the second METHODDEF.

Also, note that I copied this approach from [code.replace](https://github.com/python/cpython/blob/main/Objects/codeobject.c#L2180), so if this is not right, code.replace needs to be changed as well.